config/impl-base/src/main/java/org/jboss/arquillian/config/impl/extension/StringPropertyReplacer.java

```diff
@@ -235,47 +194,155 @@ else if (c == '}' && state == IN_BRACKET) {
         return buffer.toString();
     }
 
-    /**
-     * Try to resolve a "key" from the provided properties by
-     * checking if it is actually a "key1,key2", in which case
-     * try first "key1", then "key2". If all fails, return null.
-     * <p>
-     * It also accepts "key1," and ",key2".
-     *
-     * @param key
-     *     the key to resolve
-     * @param props
-     *     the properties to use
-     *
-     * @return the resolved key or null
-     */
-    private static String resolveCompositeKey(String key, Properties props) {
-        String value = null;
-
-        // Look for the comma
-        int comma = key.indexOf(',');
-        if (comma > -1) {
-            // If we have a first part, try resolve it
-            if (comma > 0) {
-                // Check the first part
-                String key1 = key.substring(0, comma);
-                if (props != null) {
-                    value = props.getProperty(key1);
-                } else {
-                    value = System.getProperty(key1);
+
+    public interface PropertyResolver {
+        String getValue(String key);
+    }
+
+    static class ClasspathPropertyResolver implements PropertyResolver {
```

Comment:
the problem is that it is really a hack for this specific class, it implements like a Java 8 lambda, so something to not be used outside this scope, you can find this approach in several places in JDK code. Although I agree that generally speaking is something I try always to avoid, in this case since it is for simulating lambdas, I preferred to do it in this way.

Reply:
> the problem is that it is really a hack for this specific class

But the logic is called from outside of this class - `ConfigurationSysPropResolver` & `ClasspathConfigurationPlaceholderResolver` (btw, these two classes could have similar naming format :-)). You could also call
`StringPropertyReplacer.replaceProperties(string, new ClasspathPropertyResolver())`
instead of
`StringPropertyReplacer.replaceClasspath(descriptorAsString)`

> Although I agree that generally speaking is something I try always to avoid

then don't do it :-)

>  in this case since it is for simulating lambdas, I preferred to do it in this way.

what would you say before the lamdas time? :-)

Reply:
> But the logic is called from outside of this class 

logic yes but from outside they do not know the existence of these classes. It is something internal. 

> then don't do it :-)

Generally speaking, does not mean never

Reply:
But if this is something you see bad I'll extract them outside